petastorm/tests/test_spark_dataset_converter.py

```diff
@@ -0,0 +1,245 @@
+    def test_compression(self):
+        df1 = self.spark.range(10)
+
+        converter1 = make_spark_converter(df1)
+        self.assertEqual("uncompressed",
+                         self._get_compression_type(
+                             converter1.cache_dir_url).lower())
+
+        converter2 = make_spark_converter(df1, compression=False)
+        self.assertEqual("uncompressed",
+                         self._get_compression_type(
+                             converter2.cache_dir_url).lower())
+
+        converter2 = make_spark_converter(df1, compression=True)
+        self.assertEqual("snappy",
+                         self._get_compression_type(
+                             converter2.cache_dir_url).lower())
+
+    def test_df_caching(self):
+        df1 = self.spark.range(10)
+        df2 = self.spark.range(10)
+        df3 = self.spark.range(20)
+
+        converter1 = make_spark_converter(df1)
+        converter2 = make_spark_converter(df2)
+        self.assertEqual(converter1.cache_dir_url, converter2.cache_dir_url)
+
+        converter3 = make_spark_converter(df3)
+        self.assertNotEqual(converter1.cache_dir_url,
+                            converter3.cache_dir_url)
+
+        converter11 = make_spark_converter(
+            df1, parquet_row_group_size_bytes=8 * 1024 * 1024)
+        converter21 = make_spark_converter(
+            df1, parquet_row_group_size_bytes=16 * 1024 * 1024)
+        self.assertNotEqual(converter11.cache_dir_url,
+                            converter21.cache_dir_url)
+
+        converter12 = make_spark_converter(df1, compression=True)
+        converter22 = make_spark_converter(df1, compression=False)
+        self.assertNotEqual(converter12.cache_dir_url,
+                            converter22.cache_dir_url)
+
+    def test_normalize_url(self):
+        with self.assertRaises(ValueError) as cm:
+            _normalize_dir_url('/a/b/c')
+        self.assertTrue('scheme-less' in str(cm.exception))
+
+        self.assertEqual(_normalize_dir_url('file:///a/b'), 'file:///a/b')
+        self.assertEqual(_normalize_dir_url('file:///a//b'), 'file:///a/b')
+
+    def test_df_private_caching(self):
+        self.assertTrue(_is_sub_dir_url('file:///a/b/c/d', 'file:///a/b/c'))
+        self.assertTrue(_is_sub_dir_url('hdfs:///a/b/c/d', 'hdfs:///a/b/c'))
+        self.assertTrue(_is_sub_dir_url('hdfs://nn1:9000/a/b/c/d', 'hdfs://nn1:9000/a/b/c'))
+
+        self.assertFalse(_is_sub_dir_url('file:///a/b/c', 'file:///a/b/c'))
+        self.assertFalse(_is_sub_dir_url('file:///a/b/c/d', 'file:///a/b/cc'))
+        self.assertFalse(_is_sub_dir_url('file:///a/b/c', 'file:///a/b/c/d'))
+        self.assertFalse(_is_sub_dir_url('file:///a/b/c', 'hdfs:///a/b/c'))
+        self.assertFalse(_is_sub_dir_url('hdfs://nn1:9000/a/b/c/d', 'hdfs://nn1:9001/a/b/c'))
+
+        df1 = self.spark.range(10)
+        df2 = self.spark.range(10)
+
+        converter1 = make_spark_converter(df1, cache_dir_url='file:///tmp/a1/a2')
```

Comment:
Do we expect a user to use multiple `cache_dir_url`s during one session. I am a bit confused by the subdir comparison logic. I might expect the user to be able to set the cache directory once and then not to attempt to touch it. This extra API that allows to change the cache dir in any call to `make_spark_converter` is confusing to me. Perhaps we only leave the spark-config path and read it only first time we need it? If we still leave the `cache_dir_url`, we raise an error if user tries to change the path already set.

Reply:
@mengxr WDYT ? Do we still need private cache dir for each converter ?

Reply:
The conf is local to the SparkSession, right? Then we might remove `cache_dir_url` in make_spark_converter. One scenario we need to handle is multiple users sharing the same Spark application, the conf change must be local to each user's session.

Reply:
@mengxr
Correction: 
`One scenario we need to handle is multiple users sharing the same Spark application` --> 
`One scenario we need to handle is multiple users sharing the same Spark cluster, each user launch its own spark application.`